File: hasCycle.c
/**
 * @author f4prime
 * @email zzz136454872@163.com
 * @aim a plain model for leetcode
 */

#include<stdio.h>
#include<stdlib.h>
#include<string.h>

void print_int_star(int *list, int len)
{
    for(int i=0;i<len;i++)
        printf("%d ",list[i]);
    putchar('\n');
}

void print_int_2star(int **matrix,int row,int col)
{
    for(int i=0;i<row;i++)
        print_int_star(matrix[i],col);
}

#define max(a,b) (((a)>(b))?(a):(b))

//Definition for singly-linked list.
struct ListNode {
    int val;
    struct ListNode *next;
};

bool hasCycle(struct ListNode *head) {
    if(head==NULL)
        return false;
    struct ListNode *p=head->next, *q=head;
    while(p!=NULL&&p!=q)
    {
        p=p->next;
        if(p==NULL)
            return false;
        if(p==q)
            return true;
        p=p->next;
        q=q->next;
    }
    return p==q;
}

int main()
{

    return 0;
}
    

File: hasCycle.py
from typing import Optional

from pylist import ListNode


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False
        p = head.next
        q = head

        while p != q and p is not None:
            p = p.next

            if p is None:
                return False

            if p == q:
                return True
            p = p.next
            q = q.next

        return p is not None and p == q


head = [3, 2, 0, -4]
pos = 1
head = ListNode.fromList(head)
head.next.next.next.next = head.next

head = [1, 2]
pos = 0
head = ListNode.fromList(head)
head.next.next = head

head = [1]
pos = -1
head = ListNode.fromList(head)

print(Solution().hasCycle(head))
